### PR TITLE
Add passwordless endpoints

### PR DESCRIPTION
**Pull request by lbalmaceda**
### Changes

This PR adds support for invoking the new passwordless endpoints. It also adds a small note on the readme about them and the grant type that must be enabled first.

There are some UX changes on the second request (the one that performs the authentication). The endpoint expects a `username` parameter, but I'm mapping it from either `email` or `phoneNumber`. The same happens with `otp`, which is getting mapped from `code`. Please, evaluate if this is desired or if it's only adding noise.

> Any formatting change included on the diff is due to the pre-commit hook.

### References

`SDK-1281`

### Testing

I've tested the email passwordless manually on a sample app. Unit tests are included as well.

- [x] This change adds unit test coverage
- [x] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] All existing and new tests complete without errors
- [x] All active GitHub checks have passed

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -254,6 +254,58 @@ auth0.auth
   .catch(console.error);
 ```
 
+#### Login with Passwordless
+
+Passwordless is a two-step authentication flow, and requires the **Passwordless OTP** grant to be enabled in your Auth0 application before-hand. Check [our guide](https://auth0.com/docs/dashboard/guides/applications/update-grant-types) to learn how to enable it.
```
Comment (Widcket) [inline README.md]:
```suggestion
Passwordless is a two-step authentication flow, and requires the **Passwordless OTP** grant to be enabled in your Auth0 application beforehand. Check [our guide](https://auth0.com/docs/dashboard/guides/applications/update-grant-types) to learn how to enable it.
```

File: src/auth/__tests__/index.spec.js [hunk 2]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
-        scope: 'openid'
+        scope: 'openid',
       };
       await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
     });
   });
 
+  describe('passwordless flow', () => {
+    describe('with email connection', () => {
+      it('should begin with code and authParams', async () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: You might want to have a 'begin' test just with the code and test the `authParams` separately.

File: src/auth/__tests__/index.spec.js [hunk 3]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
+        fetchMock.postOnce(
+          'https://samples.auth0.com/passwordless/start',
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: Since this url is being used twice, you might want to put it in a `const` below the nearest `describe`, or wrap that mock in a function call.

File: src/auth/__tests__/index.spec.js [hunk 4]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
+        expect.assertions(1);
+        await auth.passwordlessWithEmail({
+          email: 'info@auth0.com',
+          send: 'link',
+        });
+        expect(fetchMock.lastCall()).toMatchSnapshot();
+      });
+
+      it('should continue', async () => {
+        fetchMock.postOnce('https://samples.auth0.com/oauth/token', tokens);
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: Idem, see comment above.

File: src/auth/__tests__/index.spec.js [hunk 5]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
         state,
-        scope: 'openid'
+        scope: 'openid',
       };
       await expect(auth.exchange(parameters)).rejects.toMatchSnapshot();
     });
   });
 
+  describe('passwordless flow', () => {
+    describe('with email connection', () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: Shouldn't this be `context`?

File: src/auth/__tests__/index.spec.js [hunk 6]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
+          code: '123456',
+          audience: 'http://myapi.com',
+          scope: 'openid',
+        });
+        expect(fetchMock.lastCall()).toMatchSnapshot();
+      });
+    });
+
+    describe('with SMS connection', () => {
+      it('should begin with code and authParams', async () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: You might want to have a 'begin' test just with the code and test the authParams separately.

File: src/auth/__tests__/index.spec.js [hunk 7]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
+          email: 'info@auth0.com',
+          code: '123456',
+          audience: 'http://myapi.com',
+          scope: 'openid',
+        });
+        expect(fetchMock.lastCall()).toMatchSnapshot();
+      });
+    });
+
+    describe('with SMS connection', () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: Shouldn't this be `context`?

File: src/auth/__tests__/index.spec.js [hunk 8]
```diff
@@ -133,35 +138,143 @@ describe('auth', () => {
...
+          authParams: {
+            scope: 'openid profile',
+          },
+        });
+        expect(fetchMock.lastCall()).toMatchSnapshot();
+      });
+
+      it('should begin with link', async () => {
+        fetchMock.postOnce(
+          'https://samples.auth0.com/passwordless/start',
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: Idem as https://github.com/auth0/react-native-auth0/pull/270/files#r370146316, and all the instances below.

File: src/auth/index.js [hunk 9]
```diff
@@ -169,17 +169,147 @@ export default class Auth {
...
+      })
+      .then(responseHandler);
+  }
+
+  /**
+   * Starts the Passworldess flow with an SMS connection
+   *
+   * @param {Object} parameters passwordless parameters
+   * @param {String} parameters.phoneNumber the phone number to send the link/code to
+   * @param {String} parameters.send the passwordless strategy, either 'link' or 'code'
```
Comment (Widcket) [inline src/auth/index.js]: Isn't `link` not allowed when using SMS? See https://auth0.com/docs/api/authentication#get-code-or-link:

> You have three options for passwordless authentication:
> 
> - Send a verification code using email.
> - Send a link using email.
> - Send a verification code using SMS.

File: src/auth/index.js [hunk 10]
```diff
@@ -169,17 +169,147 @@ export default class Auth {
...
+   * @param {String} parameters.authParams optional parameters, used when strategy is 'linḱ'
```
Comment (Widcket) [inline src/auth/index.js]: Idem

File: src/auth/index.js [hunk 11]
```diff
@@ -169,17 +169,147 @@ export default class Auth {
...
+        grant_type: 'http://auth0.com/oauth/grant-type/passwordless/otp',
+      })
+      .then(responseHandler);
+  }
+
+  /**
+   * Finishes the Passworldess authentication with an SMS connection
+   *
+   * @param {Object} parameters passwordless parameters
+   * @param {String} parameters.phoneNumber the phone number where the link/code was received
```
Comment (Widcket) [inline src/auth/index.js]: Same as above.

File: src/auth/index.js [hunk 12]
```diff
@@ -169,17 +169,147 @@ export default class Auth {
...
+        },
+      },
+      parameters,
+    );
+    return this.client
+      .post('/oauth/token', {
+        ...payload,
+        client_id: this.clientId,
+        realm: 'sms',
+        grant_type: 'http://auth0.com/oauth/grant-type/passwordless/otp',
```
Comment (Widcket) [inline src/auth/index.js]: You might want to extract this, to avoid repeating it in https://github.com/auth0/react-native-auth0/pull/270/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md#diff-2b0a2893c010a4aea8147a48a8fdd092R278

File: src/auth/index.js [hunk 13]
```diff
@@ -314,20 +444,20 @@ export default class Auth {
...
-        }
+          email: {required: true},
+          password: {required: true},
+          connection: {required: true},
+          username: {required: false},
+          metadata: {required: false, toName: 'user_metadata'},
+        },
       },
-      parameters
+      parameters,
```
Comment (Widcket) [inline src/auth/index.js]: Do the style conventions for this repo allow trailing commas?

**Comment by Widcket** [comment]
> There are some UX changes on the second request (the one that performs the authentication). The endpoint expects a username parameter, but I'm mapping it from either email or phoneNumber. The same happens with otp, which is getting mapped from code. Please, evaluate if this is desired or if it's only adding noise.

I think this is good from a UX perspective.

File: src/auth/__tests__/index.spec.js (hunk 5 above)
Comment (lbalmaceda) [reply src/auth/__tests__/index.spec.js]: I don't see it being used anywhere else. It could be true for mocha tests but I don't think it's worth refactoring every test file now.

File: src/auth/__tests__/index.spec.js (hunk 3 above)
Comment (lbalmaceda) [reply src/auth/__tests__/index.spec.js]: It's a good refactor but the rest of the tests on this file at least, are doing the same.

File: src/auth/index.js (hunk 9 above)
Comment (lbalmaceda) [reply src/auth/index.js]: That's a good question actually. I don't know the answer. But on [this doc](https://auth0.com/docs/connections/passwordless/guides/sms-otp#configure-passwordless-sms-settings) they still mention link as a possible strategy for SMS connections

> Only the last one-time password **(or link)** issued will be accepted.

Let me ask the right team and come back

File: src/auth/index.js (hunk 13 above)
Comment (lbalmaceda) [reply src/auth/index.js]: Yes, enforced via pre-commit hook https://github.com/auth0/react-native-auth0/blob/master/.prettierrc.js#L5

File: src/auth/index.js (hunk 12 above)
Comment (lbalmaceda) [reply src/auth/index.js]: This particular endpoint is called for many different flows (5 calls on this file currently). What mostly changes all the time is the grant_type sent. But there are also additional parameters, such as connection/realm, and the additional values the user could include. I'm afraid that moving it into a function would maintain the same structure, as we need to pass those arguments anyway. Worth to revisit later.

File: src/auth/index.js (hunk 9 above)
Comment (lbalmaceda) [reply src/auth/index.js]: Remove this param from the doc-block.

File: src/auth/__tests__/index.spec.js [hunk 14]
```diff
@@ -133,35 +138,169 @@ describe('auth', () => {
...
+        );
+        expect.assertions(1);
+        await auth.passwordlessWithSMS({
+          phoneNumber: '+5491159991000',
+          send: 'code',
+        });
+        expect(fetchMock.lastCall()).toMatchSnapshot();
+      });
+
+      it('should begin with link', async () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js]: If we're removing the link from SMS, we should remove this test as well.

Comment (lbalmaceda) [reply src/auth/__tests__/index.spec.js]: We're not removing support for `link`. We're just not documenting it 👍 Tests should still prove the endpoint can be called with additional auth params and/or with link

File: src/auth/index.js (hunk 12 above)
Comment (Widcket) [reply src/auth/index.js]: The way I've seen that Auth0.swift solves it, is to create a custom matcher (a function, in this case) for every use case instead of every endpoint. So it has many matchers for `/oauth/token`, each one configured differently for each use case.

Comment (Widcket) [reply src/auth/index.js]: And each of these matchers may or may not take params, according to the needs of the use case.

File: README.md [hunk 15]
```diff
@@ -254,6 +254,58 @@ auth0.auth
   .catch(console.error);
 ```
 
+#### Login with Passwordless
+
+Passwordless is a two-step authentication flow, and requires the **Passwordless OTP** grant to be enabled in your Auth0 application beforehand. Check [our guide](https://auth0.com/docs/dashboard/guides/applications/update-grant-types) to learn how to enable it.
+
+To start the flow, you request a link or code to be sent to the user's email or phone number:
```
Comment (Widcket) [inline README.md]: You might want to reword this to avoid implying that you can use a link with SMS.

File: src/auth/__tests__/index.spec.js (hunk 14 above)
Comment (Widcket) [reply src/auth/__tests__/index.spec.js]: But tests are documentation, and it may be confusing for a customer to come across this.
